snapcraft/internal/meta/desktop.py

```diff
@@ -52,38 +52,39 @@ def parse_and_reformat(self, extracted_metadata: _metadata.ExtractedMetadata):
         if "Exec" not in self._parser[section]:
             raise errors.InvalidDesktopFileError(self._filename, "missing 'Exec' key")
         # XXX: do we want to allow more parameters for Exec?
-        if self._name == self._snap_name:
-            exec_value = "{} %U".format(self._name)
+        if self._app_name == self._snap_name:
+            exec_value = "{} %U".format(self._app_name)
         else:
-            exec_value = "{}.{} %U".format(self._snap_name, self._name)
+            exec_value = "{}.{} %U".format(self._snap_name, self._app_name)
         self._parser[section]["Exec"] = exec_value
         if "Icon" in self._parser[section]:
             icon = self._parser[section]["Icon"]
 
-            # Extracted metadata (e.g. from the AppStream) can override the
-            # icon location.
-            if extracted_metadata:
-                metadata_icon = extracted_metadata.get_icon()
-                if metadata_icon is not None:
-                    icon = metadata_icon
+            if icon_path is not None:
+                icon = icon_path
 
-            icon = icon.lstrip("/")
-            if icon.startswith("${SNAP}/") and os.path.exists(
-                os.path.join(self._prime_dir, icon[8:])
-            ):
-                self._parser[section]["Icon"] = icon
-            elif os.path.exists(os.path.join(self._prime_dir, icon)):
-                self._parser[section]["Icon"] = "${{SNAP}}/{}".format(icon)
-            else:
+            # Strip any leading slash.
+            icon = icon[1:] if icon.startswith("/") else icon
+            # Strip any leading ${SNAP}.
+            icon = icon[8:] if icon.startswith("${SNAP}") else icon
```

Comment:
is there a case for `$SNAP`?  This logic could neatly tuck into its own function, though I didn't see anywhere else doing the same work.

Reply:
Some predefined desktop files have this already set as it was what the recommendation was